### PR TITLE
feat: Add option to keep struct pointers nil if no sub fields are set

### PR DESCRIPTION
**Pull request by yolocs**
The existing behavior is: if a struct pointer is nil, envconfig would initiate it to be non-nil value (a pointer to the empty struct) _whether or not_ any sub-fields in that struct is set by a env var.
This PR adds the option (tag `nopreemptiveinit`) to disable that behavior. For the same case, if no sub-fields are set by env vars, the struct pointer would stay nil. _But_ if any sub-field is set by an env var, the pointer would still be initiated.

### CODE REVIEW

**Comment by sethvargo** [comment]
Hmm - this would be a breaking change. envconfig's promise is that it will deeply-initialize any nested pointers.

**Comment by yolocs** [comment]
Yea, I understand. Let's first discuss whether this behavior indeed makes sense (to me: both behaviors has its own use). Maybe we can have an option or a different func or a global flag to switch.

**Comment by sethvargo** [comment]
I'm thinking this should be a per-field configuration:

```go
type Foo struct {
  Field *Bar `env:"..., noinit"`
}
```

File: envconfig_test.go [hunk 1]
```diff
@@ -1596,6 +1596,68 @@ func TestProcessWith(t *testing.T) {
...
+			input: &struct {
+				Sub *struct {
+					Field string `env:"FIELD"`
+				} `env:",nopreemptiveinit"`
+			}{},
+			exp: &struct {
+				Sub *struct {
+					Field string `env:"FIELD"`
+				} `env:",nopreemptiveinit"`
+			}{},
```
Comment (sethvargo) [inline envconfig_test.go:1612]: Can you make this `{Sub: nil}` just so the test is more explicit?

File: envconfig_test.go [hunk 2]
```diff
@@ -1596,6 +1596,68 @@ func TestProcessWith(t *testing.T) {
...
+				},
+			},
+			// 'Sub' struct ptr should be initiated
+			// because the 'Field' is set in the lookuper.
+			lookuper: MapLookuper(map[string]string{
+				"FIELD": "banana",
+			}),
+		},
+		{
+			name: "nopreemptiveinit/no_effect_when_not_tagged",
```
Comment (sethvargo) [inline envconfig_test.go:1642]: I don't think we need this test - I think there are others that cover it?

File: envconfig_test.go [hunk 3]
```diff
@@ -1596,6 +1596,68 @@ func TestProcessWith(t *testing.T) {
...
+				}{
+					Field: "banana",
+				},
+			},
+			// 'Sub' struct ptr should be initiated
+			// because the 'Field' is set in the lookuper.
+			lookuper: MapLookuper(map[string]string{
+				"FIELD": "banana",
+			}),
+		},
```
Comment (sethvargo) [inline envconfig_test.go:1640]: Let's add a test for what happens to a non-pointer set with noinit.

```go
struct {
  Sub struct { // non-ptr
    Field string `env:"FIELD"`
  }
}
```

This is mostly to ensure we don't panic on that case.

For discussion: what should we do if someone specifies noinit on a non-pointer field? Should that be a runtime error?

File: envconfig.go [hunk 4]
```diff
@@ -217,6 +218,7 @@ type options struct {
 	Overwrite bool
 	Required  bool
 	Prefix    string
+	NoInit    bool
```
Comment (sethvargo) [inline envconfig.go:221]: I've thought about it, and I think we should have the tag also be "noinit". We have clear documentation that says "even with noinit, the struct is still initialized if a child value has an env".

**Review by sethvargo — CHANGES_REQUESTED** [review]

File: envconfig_test.go (hunk 3 above)
Comment (yolocs) [reply envconfig_test.go:1640]: I think it's better to ignore the tag if it's a non struct ptr field, like "not applicable so no effect". To make it more obvious, we could name it as `noptrinit`?

Comment (yolocs) [reply envconfig_test.go:1642]: I also update the readme saying apply the tag on non-struct-pointer fields won't have any effect.

Comment (sethvargo) [reply envconfig_test.go:1642]: I think `noinit` is the right name. I'm wondering if we should throw an error though. We already throw errors for other "impossible" configurations (e.g. specifying both _default_ and _required_)

**Pull request merged**